File: messenger/app/src/main/java/com/example/messenger/crypto/ArchiveE2EE.kt
package com.example.messenger.crypto

import android.util.Base64
import java.math.BigInteger
import java.security.AlgorithmParameters
import java.security.KeyPair
import java.security.KeyFactory
import java.security.PrivateKey
import java.security.PublicKey
import java.security.SecureRandom
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPrivateKeySpec
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec
import org.json.JSONObject

data class ArchiveIdentityState(
    val username: String,
    val publicKeyBytes: ByteArray,
    val privateKeyBytes: ByteArray,
    val version: Int = 1
) {
    fun publicKey(): PublicKey = E2EE.importPublicKey(publicKeyBytes)
    fun privateKey(): PrivateKey = E2EE.importPrivateKey(privateKeyBytes)
}

data class ArchiveKeyBundle(
    val username: String,
    val publicKeyBytes: ByteArray,
    val encryptedPrivateKey: ByteArray,
    val kdfSalt: ByteArray,
    val kdfParams: String,
    val version: Int
)

data class ArchiveEncrypted(
    val ciphertext: ByteArray,
    val nonce: ByteArray,
    val ephemeralPublicKey: ByteArray
)

object ArchiveE2EE {
    private const val KDF_ITERATIONS = 120000
    private const val NONCE_SIZE = 12
    private const val EC = "EC"
    private val P256 = ECGenParameterSpec("secp256r1")

    fun generateArchiveIdentity(username: String): ArchiveIdentityState {
        val kp: KeyPair = E2EE.generateKeyPair()
        return ArchiveIdentityState(
            username = username,
            publicKeyBytes = E2EE.exportPublicKey(kp.public),
            privateKeyBytes = E2EE.exportPrivateKey(kp.private),
            version = 1
        )
    }

    fun wrapForServer(identity: ArchiveIdentityState, password: String): ArchiveKeyBundle {
        val salt = ByteArray(16).also { SecureRandom().nextBytes(it) }
        val kek = derivePasswordKey(password, salt, KDF_ITERATIONS)
        val nonce = ByteArray(NONCE_SIZE).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, kek, GCMParameterSpec(128, nonce))
        val privB64 = Base64.encodeToString(identity.privateKeyBytes, Base64.NO_WRAP)
        val ciphertext = cipher.doFinal(privB64.toByteArray(Charsets.UTF_8))
        val combined = nonce + ciphertext
        val params = JSONObject().apply {
            put("name", "PBKDF2")
            put("hash", "SHA-256")
            put("iterations", KDF_ITERATIONS)
        }.toString()
        return ArchiveKeyBundle(
            username = identity.username,
            publicKeyBytes = identity.publicKeyBytes,
            encryptedPrivateKey = combined,
            kdfSalt = salt,
            kdfParams = params,
            version = identity.version
        )
    }

    fun restoreFromServer(
        publicKeyBytes: ByteArray,
        encryptedPrivateKey: ByteArray,
        kdfSalt: ByteArray,
        kdfParams: String,
        version: Int,
        password: String,
        username: String
    ): ArchiveIdentityState {
        val params = JSONObject(kdfParams)
        val iterations = params.optInt("iterations", KDF_ITERATIONS)
        val kek = derivePasswordKey(password, kdfSalt, iterations)
        val nonce = encryptedPrivateKey.copyOfRange(0, NONCE_SIZE)
        val ciphertext = encryptedPrivateKey.copyOfRange(NONCE_SIZE, encryptedPrivateKey.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, kek, GCMParameterSpec(128, nonce))
        val plaintext = cipher.doFinal(ciphertext)
        val decoded = String(plaintext, Charsets.UTF_8)
        val privateKeyBytes = if (decoded.trim().startsWith("{")) {
            importLegacyWebJwkPrivateKey(decoded).encoded
        } else {
            Base64.decode(decoded, Base64.NO_WRAP)
        }
        return ArchiveIdentityState(
            username = username,
            publicKeyBytes = publicKeyBytes,
            privateKeyBytes = privateKeyBytes,
            version = version
        )
    }

    fun encryptPayload(plaintext: ByteArray, recipientPublicKeyBytes: ByteArray): ArchiveEncrypted {
        val ephemeralKp = E2EE.generateKeyPair()
        val ephemeralPub = ephemeralKp.public
        val ephemeralPriv = ephemeralKp.private
        val ephemeralPubBytes = E2EE.exportPublicKey(ephemeralPub)
        val recipientPub = E2EE.importPublicKey(recipientPublicKeyBytes)
        val sharedSecret = E2EE.ecdh(ephemeralPriv, recipientPub)
        val aesKey = E2EE.hkdf(sharedSecret, info = "messenger-history-archive-v1")
        val (ciphertext, nonce) = E2EE.aesGcmEncrypt(aesKey, plaintext)
        return ArchiveEncrypted(ciphertext, nonce, ephemeralPubBytes)
    }

    fun decryptPayload(encrypted: ArchiveEncrypted, identity: ArchiveIdentityState): ByteArray {
        val ephemeralPub = E2EE.importPublicKey(encrypted.ephemeralPublicKey)
        val sharedSecret = E2EE.ecdh(identity.privateKey(), ephemeralPub)
        val aesKey = E2EE.hkdf(sharedSecret, info = "messenger-history-archive-v1")
        return E2EE.aesGcmDecrypt(aesKey, encrypted.ciphertext, encrypted.nonce)
    }

    private fun derivePasswordKey(password: String, salt: ByteArray, iterations: Int): SecretKey {
        val spec = PBEKeySpec(password.toCharArray(), salt, iterations, 256)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val keyBytes = factory.generateSecret(spec).encoded
        return SecretKeySpec(keyBytes, "AES")
    }

    private fun importLegacyWebJwkPrivateKey(json: String): PrivateKey {
        val jwk = JSONObject(json)
        require(jwk.optString("kty") == "EC") { "unsupported archive JWK type" }
        require(jwk.optString("crv") == "P-256") { "unsupported archive curve" }
        val d = base64UrlDecode(jwk.getString("d"))
        val params = AlgorithmParameters.getInstance(EC).apply { init(P256) }
            .getParameterSpec(ECParameterSpec::class.java)
        val spec = ECPrivateKeySpec(BigInteger(1, d), params)
        return KeyFactory.getInstance(EC).generatePrivate(spec)
    }

    private fun base64UrlDecode(value: String): ByteArray {
        val normalized = value
            .replace('-', '+')
            .replace('_', '/')
            .let {
                val pad = (4 - (it.length % 4)) % 4
                it + "=".repeat(pad)
            }
        return Base64.decode(normalized, Base64.DEFAULT)
    }
}

File: web/src/lib/archive-e2ee.js
const subtle = globalThis.crypto?.subtle;
const encoder = new TextEncoder();
const decoder = new TextDecoder();

function assertCrypto() {
  if (!subtle) {
    throw new Error("WebCrypto is not available");
  }
}

function randomKeyId(username, prefix = "archive") {
  return `${username}-${prefix}-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 10)}`;
}

function bytesToBase64(bytes) {
  if (typeof btoa === "function") {
    let binary = "";
    for (const byte of bytes) binary += String.fromCharCode(byte);
    return btoa(binary);
  }
  return Buffer.from(bytes).toString("base64");
}

function base64ToBytes(value) {
  if (typeof atob === "function") {
    const binary = atob(value);
    const out = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
    return out;
  }
  return new Uint8Array(Buffer.from(value, "base64"));
}

async function generateKeyMaterial() {
  const pair = await subtle.generateKey({ name: "ECDH", namedCurve: "P-256" }, true, ["deriveBits"]);
  return {
    publicKey: pair.publicKey,
    privateKey: pair.privateKey,
    publicKeyBytes: new Uint8Array(await subtle.exportKey("raw", pair.publicKey)),
    privateKeyJwk: await subtle.exportKey("jwk", pair.privateKey),
  };
}

async function importPrivateKey(privateKeyJwk) {
  return subtle.importKey("jwk", privateKeyJwk, { name: "ECDH", namedCurve: "P-256" }, true, ["deriveBits"]);
}

async function importPrivateKeyPkcs8(privateKeyBytes) {
  return subtle.importKey("pkcs8", privateKeyBytes, { name: "ECDH", namedCurve: "P-256" }, true, ["deriveBits"]);
}

async function importPublicKey(publicKeyBytes) {
  return subtle.importKey("raw", publicKeyBytes, { name: "ECDH", namedCurve: "P-256" }, false, []);
}

async function derivePasswordKey(password, salt, iterations) {
  const base = await subtle.importKey("raw", encoder.encode(password), "PBKDF2", false, ["deriveKey"]);
  return subtle.deriveKey(
    {
      name: "PBKDF2",
      hash: "SHA-256",
      salt,
      iterations,
    },
    base,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"],
  );
}

async function deriveArchiveAesKey(privateKey, peerPublicKeyBytes) {
  const peerPublicKey = await importPublicKey(peerPublicKeyBytes);
  const shared = await subtle.deriveBits({ name: "ECDH", public: peerPublicKey }, privateKey, 256);
  const hkdfKey = await subtle.importKey("raw", shared, "HKDF", false, ["deriveKey"]);
  return subtle.deriveKey(
    {
      name: "HKDF",
      hash: "SHA-256",
      salt: new Uint8Array(32),
      info: encoder.encode("messenger-history-archive-v1"),
    },
    hkdfKey,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"],
  );
}

export async function createArchiveIdentity(username) {
  assertCrypto();
  const material = await generateKeyMaterial();
  return {
    username,
    keyId: randomKeyId(username),
    algorithm: "P256-ARCHIVE-AESGCM",
    ...material,
    version: 1,
  };
}

export async function exportArchiveIdentityState(identity) {
  return {
    username: identity.username,
    keyId: identity.keyId,
    algorithm: identity.algorithm,
    version: identity.version || 1,
    publicKey: bytesToBase64(identity.publicKeyBytes),
    privateKeyJwk: identity.privateKeyJwk,
  };
}

export async function importArchiveIdentityState(state) {
  assertCrypto();
  const publicKeyBytes = base64ToBytes(state.publicKey);
  return {
    username: state.username,
    keyId: state.keyId,
    algorithm: state.algorithm,
    version: state.version || 1,
    publicKey: await importPublicKey(publicKeyBytes),
    privateKey: await importPrivateKey(state.privateKeyJwk),
    publicKeyBytes,
    privateKeyJwk: state.privateKeyJwk,
  };
}

export async function wrapArchiveIdentityForServer(identity, password) {
  assertCrypto();
  const salt = crypto.getRandomValues(new Uint8Array(16));
  const iterations = 120000;
  const kek = await derivePasswordKey(password, salt, iterations);
  const nonce = crypto.getRandomValues(new Uint8Array(12));
  const privateKeyPkcs8 = new Uint8Array(await subtle.exportKey("pkcs8", identity.privateKey));
  const ciphertext = new Uint8Array(await subtle.encrypt(
    { name: "AES-GCM", iv: nonce },
    kek,
    encoder.encode(bytesToBase64(privateKeyPkcs8)),
  ));
  const combined = new Uint8Array(nonce.length + ciphertext.length);
  combined.set(nonce, 0);
  combined.set(ciphertext, nonce.length);
  return {
    publicKey: identity.publicKeyBytes,
    encryptedPrivateKey: combined,
    kdfSalt: salt,
    kdfParams: JSON.stringify({ name: "PBKDF2", hash: "SHA-256", iterations }),
    version: identity.version || 1,
  };
}

export async function restoreArchiveIdentityFromServer(header, password, username) {
  assertCrypto();
  const params = JSON.parse(header.kdfParams || "{}");
  const encrypted = header.encryptedPrivateKey;
  const nonce = encrypted.slice(0, 12);
  const ciphertext = encrypted.slice(12);
  const kek = await derivePasswordKey(password, header.kdfSalt, Number(params.iterations || 120000));
  const plaintext = await subtle.decrypt({ name: "AES-GCM", iv: nonce }, kek, ciphertext);
  const restored = decoder.decode(plaintext);
  if (restored.trim().startsWith("{")) {
    return importArchiveIdentityState({
      username,
      keyId: `${username}-archive-restored`,
      algorithm: "P256-ARCHIVE-AESGCM",
      version: header.version || 1,
      publicKey: bytesToBase64(header.publicKey),
      privateKeyJwk: JSON.parse(restored),
    });
  }
  const privateKeyPkcs8 = base64ToBytes(restored);
  return importArchiveIdentityState({
    username,
    keyId: `${username}-archive-restored`,
    algorithm: "P256-ARCHIVE-AESGCM",
    version: header.version || 1,
    publicKey: bytesToBase64(header.publicKey),
    privateKeyJwk: await subtle.exportKey("jwk", await importPrivateKeyPkcs8(privateKeyPkcs8)),
  });
}

export async function encryptArchivePayload(plaintextBytes, recipientPublicKeyBytes) {
  assertCrypto();
  const ephemeral = await generateKeyMaterial();
  const aesKey = await deriveArchiveAesKey(ephemeral.privateKey, recipientPublicKeyBytes);
  const nonce = crypto.getRandomValues(new Uint8Array(12));
  const ciphertext = new Uint8Array(await subtle.encrypt({ name: "AES-GCM", iv: nonce }, aesKey, plaintextBytes));
  return {
    ciphertext,
    nonce,
    ephemeralPublicKey: ephemeral.publicKeyBytes,
  };
}

export async function decryptArchivePayload(payload, identity) {
  assertCrypto();
  const aesKey = await deriveArchiveAesKey(identity.privateKey, payload.ephemeralPublicKey);
  const plaintext = await subtle.decrypt({ name: "AES-GCM", iv: payload.nonce }, aesKey, payload.ciphertext);
  return new Uint8Array(plaintext);
}
